File: ios/Sources/RoadSenseNSBootstrap/Pipeline/QualityFilter.swift
import Foundation

public enum ThermalCollectionState: Equatable, Sendable {
    case nominal
    case fair
    case serious
    case critical
}

public struct DeviceCollectionState: Equatable, Sendable {
    public let thermalState: ThermalCollectionState
    public let isLowPowerModeEnabled: Bool

    public init(
        thermalState: ThermalCollectionState,
        isLowPowerModeEnabled: Bool
    ) {
        self.thermalState = thermalState
        self.isLowPowerModeEnabled = isLowPowerModeEnabled
    }
}

public enum QualityRejectionReason: Equatable, Sendable {
    case gpsAccuracy
    case speed
    case sampleCount
    case duration
    case thermal
}

public enum QualityFilterDecision: Equatable, Sendable {
    case accepted
    case rejected(QualityRejectionReason)
}

public enum QualityFilter {
    public static func evaluate(
        reading: ReadingWindow,
        deviceState: DeviceCollectionState
    ) -> QualityFilterDecision {
        if reading.gpsAccuracyMeters > 20 {
            return .rejected(.gpsAccuracy)
        }

        if reading.speedKmh < 15 || reading.speedKmh > 160 {
            return .rejected(.speed)
        }

        if reading.sampleCount < 30 {
            return .rejected(.sampleCount)
        }

        if reading.durationSeconds > 15 {
            return .rejected(.duration)
        }

        if deviceState.thermalState == .serious || deviceState.thermalState == .critical {
            return .rejected(.thermal)
        }

        return .accepted
    }
}

File: android/core-sensor/src/main/kotlin/ca/roadsense/ns/sensor/QualityFilter.kt
package ca.roadsense.ns.sensor

enum class ThermalCollectionState { NOMINAL, FAIR, SERIOUS, CRITICAL }

data class DeviceCollectionState(
    val thermalState: ThermalCollectionState,
    val isLowPowerModeEnabled: Boolean,
)

enum class QualityRejectionReason {
    GPS_ACCURACY,
    SPEED,
    SAMPLE_COUNT,
    DURATION,
    THERMAL,
}

sealed class QualityFilterDecision {
    data object Accepted : QualityFilterDecision()
    data class Rejected(val reason: QualityRejectionReason) : QualityFilterDecision()
}

object QualityFilter {
    fun evaluate(
        reading: ReadingWindow,
        deviceState: DeviceCollectionState,
    ): QualityFilterDecision {
        if (reading.gpsAccuracyMeters > 20) return QualityFilterDecision.Rejected(QualityRejectionReason.GPS_ACCURACY)
        if (reading.speedKmh < 15 || reading.speedKmh > 160) return QualityFilterDecision.Rejected(QualityRejectionReason.SPEED)
        if (reading.sampleCount < 30) return QualityFilterDecision.Rejected(QualityRejectionReason.SAMPLE_COUNT)
        if (reading.durationSeconds > 15) return QualityFilterDecision.Rejected(QualityRejectionReason.DURATION)
        if (deviceState.thermalState == ThermalCollectionState.SERIOUS ||
            deviceState.thermalState == ThermalCollectionState.CRITICAL
        ) {
            return QualityFilterDecision.Rejected(QualityRejectionReason.THERMAL)
        }
        return QualityFilterDecision.Accepted
    }
}
